package-name-map/src/package-name-map.ts

```diff
@@ -0,0 +1,248 @@
+    // the Package's main file, otherwise remove the package name from the
+    // specifier to get the package-internal path to the module.
+    const modulePath =
+      specifier === packageName
+        ? pkg.main
+        : specifier.substring(packageName!.length);
+
+    return joinPaths(
+      packagePathPrefix,
+      packageScope.path_prefix,
+      pkg.path || packageName,
+      modulePath
+    );
+  }
+}
+
+/**
+ * Returns the appliciable Scope for `referrerURL`, and its ancestors.
+ *
+ * TODO: A recursive algorithm might be easier to write spec text for.
+ */
+const getScopeContext = (rootScope: Scope, referrerURL: string) => {
+  const scopeContext: ScopeEntry[] = [];
+
+  let currentScope: Scope = rootScope;
+  let currentScopePath = '';
+  let currentScopePathPrefix = '';
+  let foundChildScope = true;
+
+  while (foundChildScope) {
+    scopeContext.push({path: currentScopePath, scope: currentScope});
+    foundChildScope = false;
+    if (currentScope.scopes !== undefined) {
+      for (const [childScopePrefix, childScope] of Object.entries(
+        currentScope.scopes
+      )) {
+        const childScopeFullPrefix = joinPaths(
+          currentScopePathPrefix,
+          childScopePrefix
+        );
+        if (isPathSegmentPrefix(childScopeFullPrefix, referrerURL)) {
+          currentScope = childScope;
+          currentScopePath = childScopePrefix;
+          currentScopePathPrefix = childScopeFullPrefix;
+          foundChildScope = true;
+          break;
+        }
+      }
+    }
+  }
+  return scopeContext;
+};
+
+/**
+ * Given a list of Scopes (ancestors from a root Scope to the target search
+ * Scope) to search, finds a Package whose name is a PathSegmentPrefix of
+ * `specifier`.
+ *
+ * Returns:
+ *  - A sublist of the given Scope list that contains the scope that the package
+ *    entry was found in, and its ancestors.
+ *  - The found package
+ *  - The found package name
+ *
+ * TODO: A recursive algorithm might be easier to write spec text for.
+ */
+const findPackage = (
+  scopes: ScopeEntry[],
+  specifier: string
+): FindPackageResult => {
+  let scopePathIndex = scopes.length - 1;
+  let foundPackage: Package | undefined = undefined;
+  let foundPackageName: string | undefined = undefined;
+
+  do {
+    const scope = scopes[scopePathIndex].scope;
+    if (scope.packages !== undefined) {
+      for (const [pkgName, pkg] of Object.entries(scope.packages)) {
+        if (isPathSegmentPrefix(pkgName, specifier)) {
+          foundPackage = pkg;
+          foundPackageName = pkgName;
+          break;
+        }
+      }
+    }
+  } while (foundPackage === undefined && --scopePathIndex >= 0);
+  return {
+    scopeContext: scopes.slice(0, scopePathIndex + 1),
+    packageName: foundPackageName,
+    package: foundPackage,
+  };
+};
+
+/**
+ * Joins together two or more strings, ensuring there's a '/' character
+ * between each.
+ */
+const joinPaths = (...paths: Array<string | undefined>) => {
```

Comment:
Hmm why is `undefined` allowed in this array? (Makes it harder to write spec text for.)

Reply:
Removed and redefined this function to perform successive URL resolution instead of path-joining.